README.md
# MultipleTesting

What is better than one statistical test? Many tests!

`MultipleTesting` offers commonly used algorithms to handle situations when dealing with multiple hypothesis tests – well-tested and efficiently implemented for large-scale analyses.

Features:
- [Adjustment of p-Values](#adjustment-of-p-values)
- [Estimation of π₀](#estimation-of-π₀)
- [Combination of p-Values](#combination-of-p-values)
- [Higher Criticism](#higher-criticism)
- [Modelling of p-Value Distributions](#modelling-of-p-value-distributions)


## Package Status

[![Documentation](https://img.shields.io/badge/docs-stable-blue.svg)](https://juliangehring.github.io/MultipleTesting.jl/stable) [![DOI](https://zenodo.org/badge/27935122.svg)](https://zenodo.org/badge/latestdoi/27935122)

![Testing](https://github.com/juliangehring/MultipleTesting.jl/workflows/Testing/badge.svg) [![Coverage](https://codecov.io/gh/juliangehring/MultipleTesting.jl/branch/master/graph/badge.svg)](https://codecov.io/gh/juliangehring/MultipleTesting.jl)


## Features


### Adjustment of p-Values

```julia
adjust(pvalues, <:PValueAdjustmentMethod)
```

The adjustment can also be performed on the `k` smallest out of `n` p-values:

```julia
adjust(pvalues, n, <:PValueAdjustmentMethod)
```


#### Bonferroni
    
```julia
adjust(pvalues, Bonferroni())
```

Bonferroni, C.E. (1936). Teoria statistica delle classi e calcolo delle probabilita
(Libreria internazionale Seeber).


#### Benjamini-Hochberg

```julia
adjust(pvalues, BenjaminiHochberg())
```

Adaptive Benjamini-Hochberg with known π₀ or π₀ estimation method.

```julia
adjust(pvalues, BenjaminiHochbergAdaptive(π₀))
adjust(pvalues, BenjaminiHochbergAdaptive(<:PValueAdjustmentMethod))
```

Benjamini, Y., and Hochberg, Y. (1995). Controlling the False Discovery Rate: A
Practical and Powerful Approach to Multiple Testing. Journal of the Royal
Statistical Society. Series B (Methodological) 57, 289–300.

Benjamini, Y., Krieger, A. M. & Yekutieli, D. (2006). Adaptive linear step-up 
procedures that control the false discovery rate. Biometrika 93, 491–507.


#### Benjamini-Yekutieli

```julia
adjust(pvalues, BenjaminiYekutieli())
```

Benjamini, Y., and Yekutieli, D. (2001). The Control of the False Discovery Rate
in Multiple Testing under Dependency. The Annals of Statistics 29, 1165–1188.


#### Benjamini-Liu

```julia
adjust(pvalues, BenjaminiLiu())
```

Benjamini, Y., and Liu, W. (1999). A step-down multiple hypotheses testing
procedure that controls the false discovery rate under independence. Journal of
Statistical Planning and Inference 82, 163–170.


#### Hochberg

```julia
adjust(pvalues, Hochberg())
```

Hochberg, Y. (1988). A sharper Bonferroni procedure for multiple tests of
significance. Biometrika 75, 800–802.


#### Holm

```julia
adjust(pvalues, Holm())
```

Holm, S. (1979). A Simple Sequentially Rejective Multiple Test Procedure.
Scandinavian Journal of Statistics 6, 65–70.


#### Hommel

```julia
adjust(pvalues, Hommel())
```

Hommel, G. (1988). A stagewise rejective multiple test procedure based on a
modified Bonferroni test. Biometrika 75, 383–386.


#### Sidak

```julia
adjust(pvalues, Sidak())
```

Šidák, Z. (1967). Rectangular Confidence Regions for the Means of Multivariate
Normal Distributions. Journal of the American Statistical Association 62,
626–633.


#### Forward Stop

```julia
adjust(pvalues, ForwardStop())
```

G’Sell, M.G., Wager, S., Chouldechova, A., and Tibshirani, R. (2016). Sequential
selection procedures and false discovery rate control. J. R. Stat. Soc. B 78,
423–444.


#### Barber-Candès

```julia
adjust(pvalues, BarberCandes())
```

Barber, R.F., and Candès, E.J. (2015). Controlling the false discovery rate via
knockoffs. Ann. Statist. 43, 2055–2085.

Arias-Castro, E., and Chen, S. (2017). Distribution-free multiple testing.
Electron. J. Statist. 11, 1983–2001.



### Estimation of π₀

```julia
estimate(pvalues, <:Pi0Estimator)
```


#### Storey

```julia
estimate(pvalues, Storey())
```

Storey, J.D., Taylor, J.E., and Siegmund, D. (2004). Strong control,
conservative point estimation and simultaneous conservative consistency of false
discovery rates: a unified approach. Journal of the Royal Statistical Society:
Series B (Statistical Methodology) 66, 187–205.


#### Storey's Closed-Form Bootstrap

```julia
estimate(pvalues, StoreyBootstrap())
```

Robinson, D. (2016). Original Procedure for Choosing λ.
http://varianceexplained.org/files/pi0boot.pdf


#### Least Slope

```julia
estimate(pvalues, LeastSlope())
```

Benjamini, Y., and Hochberg, Y. (2000). On the Adaptive Control of the False
Discovery Rate in Multiple Testing With Independent Statistics. Journal of
Educational and Behavioral Statistics 25, 60–83.


#### Two Step

```julia
estimate(pvalues, TwoStep())
estimate(pvalues, TwoStep(α))
estimate(pvalues, TwoStep(α, <:PValueAdjustmentMethod)
```

Benjamini, Y., Krieger, A.M., and Yekutieli, D. (2006). Adaptive linear step-up
procedures that control the false discovery rate. Biometrika 93, 491–507.


#### Right Boundary

Storey's estimate with dynamically chosen λ

```julia
estimate(pvalues, RightBoundary())
```

Liang, K., and Nettleton, D. (2012). Adaptive and dynamic adaptive procedures
for false discovery rate control and estimation. Journal of the Royal
Statistical Society: Series B (Statistical Methodology) 74, 163–182.


#### Beta-Uniform Mixture (BUM)

```julia
estimate(pvalues, BUM())
```

Pounds, S., and Morris, S.W. (2003). Estimating the occurrence of false
positives and false negatives in microarray studies by approximating and
partitioning the empirical distribution of p-values. Bioinformatics 19,
1236–1242.


#### Censored Beta-Uniform Mixture

```julia
estimate(pvalues, CensoredBUM())
```

Markitsis, A., and Lai, Y. (2010). A censored beta mixture model for the
estimation of the proportion of non-differentially expressed genes.
Bioinformatics 26, 640–646.


#### Flat Grenander

```julia
estimate(pvalues, FlatGrenander())
```

Langaas, M., Lindqvist, B.H., and Ferkingstad, E. (2005). Estimating the
proportion of true null hypotheses, with application to DNA microarray data.
Journal of the Royal Statistical Society: Series B (Statistical Methodology) 67,
555–572.


#### Convex Decreasing

```julia
estimate(pvalues, ConvexDecreasing())
fit(ConvexDecreasing(), pvalues)
```

Langaas, M., Lindqvist, B.H., and Ferkingstad, E. (2005). Estimating the
proportion of true null hypotheses, with application to DNA microarray data.
Journal of the Royal Statistical Society: Series B (Statistical Methodology) 67,
555–572.


#### Oracle for Known π₀

```julia
estimate(pvalues, Oracle(π₀))
```



### Combination of p-Values

```julia
combine(pvalues, <:PValueCombination)
```


#### Fisher

```julia
combine(pvalues, Fisher())
```

Fisher, R.A. (1925). Statistical methods for research workers (Genesis
Publishing Pvt Ltd).


#### Stouffer

Optionally with weights

```julia
combine(pvalues, Stouffer())
combine(pvalues, weights, Stouffer())
```

Stouffer, S.A. (1949). The American soldier. Vol. 1: Adjustment during army life
(Princeton University Press).

Liptak, T. (1958). On the combination of independent tests. Magyar Tud Akad Mat
Kutato Int Kozl 3, 171–197.


#### Logit

```julia
combine(pvalues, Logit())
```

Mudholkar, G.S., and George, E.O. (1977). The Logit Statistic for Combining
Probabilities - An Overview (Rochester University NY, Dept of Statistics).


#### Tippett

```julia
combine(pvalues, Tippett())
```

Tippett, L.H.C. (1931). The Methods of Statistics. An introduction mainly for
workers in the biological sciences.


#### Simes

```julia
combine(pvalues, Simes())
```

Simes, R.J. (1986). An improved Bonferroni procedure for multiple tests of
significance. Biometrika 73, 751–754.


#### Wilkinson

```julia
combine(pvalues, Wilkinson(rank))
```

Wilkinson, B. (1951). A statistical consideration in psychological research.
Psychological Bulletin 48, 156.


#### Minimum of Adjusted p-Values

```julia
combine(pvalues, Minimum(PValueAdjustment()))
```



### Higher Criticism

Higher criticism scores and threshold

```julia
estimate(pvalues, HigherCriticismScores())
estimate(pvalues, HigherCriticismThreshold())
```

Donoho, D., and Jin, J. (2008). Higher criticism thresholding: Optimal feature
selection when useful features are rare and weak. PNAS 105, 14790–14795.

Klaus, B., and Strimmer, K. (2013). Signal identification for rare and weak
features: higher criticism or false discovery rates? Biostatistics 14, 129–143.



### Modelling of p-Value Distributions


#### Beta Uniform Mixture Model

```julia
BetaUniformMixtureModel(π₀, α, β)
```

Pounds, S., and Morris, S.W. (2003). Estimating the occurrence of false
positives and false negatives in microarray studies by approximating and
partitioning the empirical distribution of p-values. Bioinformatics 19,
1236–1242.


## Installation

The `MultipleTesting` package is part of the Julia ecosphere and the latest
release version can be installed with

```julia
pkg> add MultipleTesting
```

More details on packages and how to manage them can be found in the
[package section](https://docs.julialang.org/en/v1/stdlib/Pkg/index.html)
of the Julia documentation.


NEWS.md
# MultipleTesting.jl News and Changes

## Version 1.0.0

### Changes

- Release stable version

### Support

- Supports julia v1.6 and newer


## Version 0.6.0

### Changes

- Upgrades julia compat to 1.6 (#127)
- Upgrade StatsBase compat to 0.34 (#127)
- Fixes tests where the error type changed in recent julia versions (#127)

### Support

- Supports julia v1.6 and newer


## Version 0.5.1

### Changes

- Handle dependencies automatically through CompatHelper
- Extend support to SpecialFunctions v2
- Update the readme


## Version 0.5.0

### Changes

- Improve the reordering strategy of p-values in adjustment methods. This change saves one sorting step for all adjustment methods that require sorted p-values. As a result, the performance for these methods is significantly improved.
- Rename p-value combination types, in order to improve consistency with other types.


### Support

- Supports julia v1.4 and newer


## Version 0.4.1

### Changes

- Update continuous integration configuration
- Add support for julia v1.0

### Support

- Supports julia v0.7 and v1.0


## Version 0.4.0

### Changes

#### User-facing changes

- Require julia v0.7 as the minimum version, drop support for julia v0.6 
- Describe package features in the documentation


### Support

- Requires julia 0.7


## Version 0.3.1

### New Features

- Build and deploy the package documentation (#94)


### Changes

#### User-facing changes

- Fix Barber-Candes adjustemt if all p-values are < 0.5 (#93)
- Update the documentation notebooks (#88)
- Type the `adjust` methods (#90)
- Rename the adjustment field in `TwoStep` type (#89)


#### Internal changes

- Optimise the Hommel adjustment (#91)
- Enable fast finishing for CI builds (#92)


## Version 0.3.0

### New Features

- Convex decreasing density estimator for π0 (#56)
- Higher criticism scores (`HigherCriticismScores`) and threshold (`HigherCriticismThreshold`) estimation (#55)
- List of publications which describe the implemented statistical methods (#54, #65, #81, #83)


### Changes

#### User-facing changes

- Fix p-value adjustments with unsorted inputs (#66)
- Speed up p-value combinations (#75)
- Define return types for p-value combinations (#76)
- Unify methods for weighted Stouffer combination (#84)
- Update documentation notebooks (#67)
- Drop support for julia 0.5, require julia 0.6 as the minimum version (#52, #64, #58, #82)


#### Internal changes

- Simplify the step functions for p-value adjustments (#77)
- Simplify the method structure (#74)
- Switch to new julia syntax, with preliminary support for julia 0.7 (#71, #78)
- Update tests and references for p-value combinations (#70)
- Use `clamp` for better bounding of estimates (#85, #82)
- Support the new `Distributions` interface (#52)
- Drop the `Compat` dependency (#53)
- Link comic logo in the readme file as an external resource (#69)


### Removed

- Drop `qvalues` in favour of BH-adjusted p-values (#73).


### Support

- Requires julia 0.6
- Preliminary support for julia 0.7 nightly builds


## Version 0.2.3

### Changes

- Switch to new `Weights` type in 'StatsBase' (>= v0.15.0)


## Version 0.2.2

### Changes

- Makes the `PValues` type immutable and persistent
- Finalises support of julia 0.6 with testing


### Support

- Requires julia 0.5 or 0.6


## Version 0.2.1

### Changes

- Adds compatibility with julia 0.6 development versions, including automated tests
- Fixes Censored BUM π₀ edge case when all p-values are equal to 1
- Updates test cases and reference values for `qValues`
- Introduces `Compat` as a new dependency
- Separates test coverage reports for different julia versions


### Support

- Requires julia 0.5 or 0.6


## Version 0.2.0

### New Features

- Combination of p-values through the `combine` interface:
  * Fisher
  * Stouffer
  * Logit
  * Wilkinson
  * Simes
  * Tippett
  * Minimum of adjusted p-values
- Analysis notebook for p-value combinations
- `PValues` type: Representation of vectors with p-values
- Adjustments of p-values support specifying the total number of tests
- Barber-Candès p-value adjustment


### Changes

- Optimised Benjamini-Liu p-value adjustment and test cases
- Estimator types are now immutables
- Harmonised type names
- Restructed readme
- Simplified method signatures
- Various performance improvements


### Removed

- Export of lower-level functions for p-value adjustment


### Support

- Requires julia 0.5


## Version 0.1.0

### New Features

- New π₀ estimators: BUM, Censored BUM, Flat Grenander
- New p-value adjustment methods: Forward Stop, Benjamini-Liu
- BUM model for simulating p-value distributions
- Fast isotonic regression
- Reference list of related software packages in documentation


### Changes

- Grenander ECDF estimator
- Immutable method types
- Default α value for Two Step constructor
- Tests use new julia 0.5 base testing framework
- Documentation moved to 'docs' directory


### Removed

- Test coverage reporting through coveralls


### Support

- Requires julia 0.5


## Version 0.0.2


### New features

- New π₀ estimators: Two Step, Right Boundary
- New p-value adjustment methods: Benjamini-Hochberg Adaptive, Oracle, Sidak
- Package precompilation


### Changes / Changed

- Redefined Oracle Benjamini-Hochberg as Adaptive Benjamini-Hochberg procedure


### Support

- Requires julia 0.4


src/pi0-estimators.jl
### estimators for π₀

"""
    estimate(PValues, <:Pi0Estimator)

Estimate π₀, the fraction of tests under the null hypothesis


# Examples

```jldoctest
julia> pvals = PValues([0.001, 0.002, 0.01, 0.03, 0.5]);

julia> estimate(pvals, StoreyBootstrap())
0.0
julia> estimate(pvals, FlatGrenander())
0.42553191489361697
```


# See also

`Pi0Estimator`s:

[`Storey`](@ref)
[`StoreyBootstrap`](@ref)
[`LeastSlope`](@ref)
[`Oracle`](@ref)
[`TwoStep`](@ref)
[`RightBoundary`](@ref)
[`CensoredBUM`](@ref)
[`BUM`](@ref)
[`FlatGrenander`](@ref)
[`ConvexDecreasing`](@ref)

"""
function estimate end

function estimate(pValues::AbstractVector{T}, method::M) where {T <: AbstractFloat,M <: Pi0Estimator}
    return estimate(PValues(pValues), method)
end


## Storey estimator

"""
Storey's π₀ estimator


# Examples

```jldoctest
julia> pvals = PValues([0.001, 0.002, 0.01, 0.03, 0.5]);

julia> estimate(pvals, Storey())
0.22222222222222224

julia> estimate(pvals, Storey(0.4))
0.33333333333333337
```


# References

Storey, J.D., Taylor, J.E., and Siegmund, D. (2004). Strong control,
conservative point estimation and simultaneous conservative consistency of false
discovery rates: a unified approach. Journal of the Royal Statistical Society:
Series B (Statistical Methodology) 66, 187–205.

"""
struct Storey <: Pi0Estimator
    λ::Float64

    function Storey(λ)
        isin(λ, 0, 1) || throw(DomainError("λ must be in [0, 1]"))
        return new(λ)
    end
end

Storey() = Storey(0.1)

function estimate(pValues::PValues{T}, pi0estimator::Storey) where T <: AbstractFloat
    lambda = pi0estimator.λ
    pi0 = (sum(pValues .>= lambda) / length(pValues)) / (1 - lambda)
    pi0 = clamp(pi0, 0, 1)
    return pi0
end


## Storey bootstrap estimator

"""
Storey's closed-form bootstrap π₀ estimator


# Examples

```jldoctest
julia> pvals = PValues([0.001, 0.002, 0.01, 0.03, 0.5]);

julia> estimate(pvals, StoreyBootstrap())
0.0

julia> estimate(pvals, StoreyBootstrap(0.1:0.1:0.9, 0.2))
0.0
```


# References

Robinson, D. (2016). Original Procedure for Choosing λ.
http://varianceexplained.org/files/pi0boot.pdf

Storey, J.D., Taylor, J.E., and Siegmund, D. (2004). Strong control,
conservative point estimation and simultaneous conservative consistency of false
discovery rates: a unified approach. Journal of the Royal Statistical Society:
Series B (Statistical Methodology) 66, 187–205.

"""
struct StoreyBootstrap <: Pi0Estimator
    λseq::Vector{Float64}
    q::Float64

    function StoreyBootstrap(λseq, q)
        isin(λseq, 0, 1) || throw(DomainError("λseq must be in [0, 1]"))
        isin(q, 0, 1) || throw(DomainError("q must be in [0, 1]"))
        return new(λseq, q)
    end
end

StoreyBootstrap() = StoreyBootstrap(0.05:0.05:0.95, 0.1)

function estimate(pValues::PValues{T}, pi0estimator::StoreyBootstrap) where T <: AbstractFloat
    lambdas = pi0estimator.λseq
    q = pi0estimator.q
    n = length(pValues)
    w = [sum(pValues .>= l) for l in lambdas]
    pi0 = w ./ n ./ (1 .- lambdas)
    min_pi0 = quantile(pi0, q)
    mse = (w ./ (n.^2 .* (1 .- lambdas).^2 )) .* (1 .- w / n) + (pi0 .- min_pi0).^2
    pi0 = clamp(pi0[argmin(mse)], 0, 1)
    return pi0
end


## Least SLope (LSL) estimator

"""
Least Slope (LSL) π₀ estimator


# Examples

```jldoctest
julia> pvals = PValues([0.001, 0.002, 0.01, 0.03, 0.5]);

julia> estimate(pvals, LeastSlope())
1.0
```


# References

Benjamini, Y., and Hochberg, Y. (2000). On the Adaptive Control of the False
Discovery Rate in Multiple Testing With Independent Statistics. Journal of
Educational and Behavioral Statistics 25, 60–83.

"""
struct LeastSlope <: Pi0Estimator
end

function estimate(pValues::PValues{T}, pi0estimator::LeastSlope) where T <: AbstractFloat
    n = length(pValues)
    pValues = sort(pValues)
    s0 = lsl_slope(1, n, pValues)
    sx = 0
    for i in 2:n
        s1 = lsl_slope(i, n, pValues)
        if (s1 - s0) < 0
            sx = s1
            break
        end
        s0 = s1
    end
    pi0 = min(1 / sx + 1, n) / n
    return pi0
end

function lsl_slope(i::Integer, n::Integer, pValue::AbstractVector{T}) where T <: AbstractFloat
    s = (1 - pValue[i]) / (n - i + 1)
    return s
end


## Oracle

"""
Oracle π₀


# Examples

```jldoctest
julia> pvals = PValues([0.001, 0.002, 0.01, 0.03, 0.5]);

julia> estimate(pvals, Oracle(0.5)) # a bit boring...
0.5
```
"""
struct Oracle <: Pi0Estimator
    π0::Float64

    function Oracle(π0)
        isin(π0, 0, 1) || throw(DomainError("π0 must be in [0, 1]"))
        return new(π0)
    end
end

Oracle() = Oracle(1.0)

function estimate(pValues::PValues{T}, pi0estimator::Oracle) where T <: AbstractFloat
    return pi0estimator.π0
end


## Two-Step estimator: Benjamini, Krieger and Yekutieli (2006)

"""
Two-step π₀ estimator


# Examples

```jldoctest
julia> pvals = PValues([0.001, 0.002, 0.01, 0.03, 0.5]);

julia> estimate(pvals, TwoStep())
0.2

julia> estimate(pvals, TwoStep(0.05, BenjaminiLiu()))
0.2

```


# References

Benjamini, Y., Krieger, A.M., and Yekutieli, D. (2006). Adaptive linear step-up
procedures that control the false discovery rate. Biometrika 93, 491–507.

"""
struct TwoStep <: Pi0Estimator
    α::Float64
    adjustment::PValueAdjustment

    function TwoStep(α, method)
        isin(α, 0, 1) || throw(DomainError("α must be in [0, 1]"))
        return new(α, method)
    end
end

TwoStep() = TwoStep(0.05)

TwoStep(α) = TwoStep(α, BenjaminiHochberg())

function estimate(pValues::PValues{T}, pi0estimator::TwoStep) where T <: AbstractFloat
    alpha = pi0estimator.α
    padj = adjust(pValues, pi0estimator.adjustment)
    pi0 = sum(padj .>= (alpha / (1 + alpha))) / length(padj)
    return pi0
end


# RightBoundary procedure as defined in Definition 2 of Liang and Nettleton, 2012

"""
Right boundary π₀ estimator


# Examples

```jldoctest
julia> pvals = PValues([0.001, 0.002, 0.01, 0.03, 0.5]);

julia> estimate(pvals, RightBoundary())
0.2127659574468085

julia> estimate(pvals, RightBoundary(0.1:0.1:0.9))
0.25
```


# References

Liang, K., and Nettleton, D. (2012). Adaptive and dynamic adaptive procedures
for false discovery rate control and estimation. Journal of the Royal
Statistical Society: Series B (Statistical Methodology) 74, 163–182.

"""
struct RightBoundary <: Pi0Estimator
    λseq::Vector{Float64}

    function RightBoundary(λseq)
        isin(λseq, 0, 1) || throw(DomainError("λseq must be in [0, 1]"))
        return new(λseq)
    end
end

# λseq used in Liang, Nettleton 2012
RightBoundary() = RightBoundary([0.02:0.02:0.1; 0.15:0.05:0.95])

function estimate(pValues::PValues{T}, pi0estimator::RightBoundary) where T <: AbstractFloat
    n = length(pValues)
    λseq = sort(pi0estimator.λseq)
    # make sure we catch p-values equal to 1 despite left closure
    # use closed=:left because we have been using >= convention in this package
    # note that original paper uses > convention.
    h = fit(Histogram, pValues, [λseq; Inf], closed = :left)
    pi0_estimates = reverse(cumsum(reverse(h.weights))) ./ (1 .- λseq) ./ n
    pi0_decrease = diff(pi0_estimates) .>= 0
    pi0_decrease[end] = true
    pi0 = pi0_estimates[something(findfirst(pi0_decrease), 0) + 1]
    pi0 = clamp(pi0, 0, 1)
    return pi0
end


## Censored BUM

"""
Censored Beta-Uniform Mixture (censored BUM) π₀ estimator


# Examples

```jldoctest
julia> pvals = PValues([0.001, 0.002, 0.01, 0.03, 0.5]);

julia> estimate(pvals, CensoredBUM())
0.21052495526400936

```


# References

Markitsis, A., and Lai, Y. (2010). A censored beta mixture model for the
estimation of the proportion of non-differentially expressed genes.
Bioinformatics 26, 640–646.

"""
struct CensoredBUM <: Pi0Estimator
    γ0::Float64
    λ::Float64
    xtol::Float64
    maxiter::Int64

    function CensoredBUM(γ0, λ, xtol, maxiter)
        isin(γ0, 0, 1) || throw(DomainError("γ0 must be in [0, 1]"))
        isin(λ, 0, 1) || throw(DomainError("λ must be in [0, 1]"))
        isin(xtol, 0, 1) || throw(DomainError("xtol must be in [0, 1]"))
        maxiter > 0 || throw(DomainError("maxiter must be a positive number"))
        return new(γ0, λ, xtol, maxiter)
    end
end

CensoredBUM() = CensoredBUM(0.5, 0.05, 1e-6, 10000)
CensoredBUM(γ0, λ) = CensoredBUM(γ0, λ, 1e-6, 10000)
    
struct CensoredBUMFit <: Pi0Fit
    π0::Float64
    param::Vector{Float64}
    is_converged::Bool
end

function fit(pi0estimator::CensoredBUM, pValues::AbstractVector{T}; kw...) where T <: AbstractFloat
    π0, param, is_converged = cbum_pi0(pValues, pi0estimator.γ0, pi0estimator.λ,
                                       pi0estimator.xtol, pi0estimator.maxiter;
                                       kw...)
    return CensoredBUMFit(π0, param, is_converged)
end

function estimate(pValues::PValues{T}, pi0estimator::CensoredBUM) where T <: AbstractFloat
    return estimate(fit(pi0estimator, pValues))
end

function estimate(pi0fit::CensoredBUMFit)
    pi0 = pi0fit.is_converged ? pi0fit.π0 : NaN
    return pi0
end

function cbum_pi0(pValues::AbstractVector{T},
                  γ0::AbstractFloat = 0.5, λ::AbstractFloat = 0.05,
                  xtol::AbstractFloat = 1e-6, maxiter::Integer = 10000) where T <: AbstractFloat
    n = length(pValues)
    idx_right = pValues .>= λ
    n2 = sum(idx_right)
    n1 = n - n2
    sz = (1 - γ0) * n
    szr = (1 - γ0) * n2
    szl = sz - szr
    pr = pValues[idx_right]
    lpr = log.(pr)
    ll = log(λ)
    zr = fill(1 - γ0, n2)
    pi0_old = γ0 = α = γ = Inf
    for i in 1:maxiter
        γ = 1 - sz / n
        α = -szr / ( ll * szl + sum(zr .* lpr) )
        γ = clamp(γ, 0, 1)
        α = clamp(α, 0, 1)
        xl = (1 - γ) * (λ^α)
        szl = (xl ./ (γ * λ + xl)) * n1
        xr = (1 - γ) * α * pr.^(α - 1)
        zr = xr ./ (γ .+ xr)
        szr = sum(zr)
        sz = szl + szr
        pi0_new = γ + (1 - γ) * α
        if abs(pi0_new - pi0_old) <= xtol
            return pi0_new, [γ, α], true
        end
        γ0 = γ
        pi0_old = pi0_new
    end
    return NaN, [γ, α], false
end


## BUM

"""
Beta-Uniform Mixture (BUM) π₀ estimator


# Examples

```jldoctest
julia> pvals = PValues([0.001, 0.002, 0.01, 0.03, 0.5]);

julia> estimate(pvals, BUM())
0.22802795505154264
```


# References

Pounds, S., and Morris, S.W. (2003). Estimating the occurrence of false
positives and false negatives in microarray studies by approximating and
partitioning the empirical distribution of p-values. Bioinformatics 19,
1236–1242.

"""
struct BUM <: Pi0Estimator
    γ0::Float64
    xtol::Float64
    maxiter::Int64

    function BUM(γ0, xtol, maxiter)
        isin(γ0, 0, 1) || throw(DomainError("γ0 must be in [0, 1]"))
        isin(xtol, 0, 1) || throw(DomainError("xtol must be in [0, 1]"))
        maxiter > 0 || throw(DomainError("maxiter must be a positive number"))
        return new(γ0, xtol, maxiter)
    end
end

BUM() = BUM(0.5, 1e-6, 10000)

BUM(y0::Float64) = BUM(y0, 1e-6, 10000)

struct BUMFit <: Pi0Fit
    π0::Float64
    param::Vector{Float64}
    is_converged::Bool
end

function fit(pi0estimator::BUM, pValues::AbstractVector{T}; kw...) where T <: AbstractFloat
    π0, param, is_converged = cbum_pi0(pValues, pi0estimator.γ0, eps(),
                                       pi0estimator.xtol, pi0estimator.maxiter;
                                       kw...)
    return BUMFit(π0, param, is_converged)
end

function estimate(pValues::PValues{T}, pi0estimator::BUM) where T <: AbstractFloat
    return estimate(fit(pi0estimator, pValues))
end

function estimate(pi0fit::BUMFit)
    pi0 = pi0fit.is_converged ? pi0fit.π0 : NaN
    return pi0
end


## Longest constant interval in the Grenander estimator

"""
Flat Grenander π₀ estimator

Estimates π₀ by finding the longest constant interval in the Grenander estimator.


# Examples

```jldoctest
julia> pvals = PValues([0.001, 0.002, 0.01, 0.03, 0.5]);

julia> estimate(pvals, FlatGrenander())
0.42553191489361697
```


# References

Langaas, M., Lindqvist, B.H., and Ferkingstad, E. (2005). Estimating the
proportion of true null hypotheses, with application to DNA microarray data.
Journal of the Royal Statistical Society: Series B (Statistical Methodology) 67,
555–572.

"""
struct FlatGrenander <: Pi0Estimator
end

function estimate(pValues::PValues{T}, pi0estimator::FlatGrenander) where T <: AbstractFloat
    p, f, F = grenander(pValues)
    pi0 = longest_constant_interval(p, f)
    return pi0
end

function longest_constant_interval(p::AbstractVector{T}, f::AbstractVector{T}) where T <: AbstractFloat
    p = [0.0; p]
    f = [Inf; f]
    i2 = length(f)
    Δp_max = Δp = 0.0
    pi0 = 1
    for i1 in (length(f) - 1):-1:1
        if f[i2] ≈ f[i1] # within constant interval
            Δp = p[i2] - p[i1]
        else
            if Δp >= Δp_max
                Δp_max = Δp
                pi0 = f[i2]
            end
            i2 = i1
            Δp = 0.0
        end
        if f[i1] > 1
            break
        end
    end
    return pi0
end


## Convex Decreasing π₀ estimator

"""
Convex Decreasing π₀ estimator


# Examples

```jldoctest
julia> pvals = PValues([0.001, 0.002, 0.01, 0.03, 0.5]);

julia> estimate(pvals, ConvexDecreasing())
0.013007051336745304
```


# References

Langaas, M., Lindqvist, B.H., and Ferkingstad, E. (2005). Estimating the
proportion of true null hypotheses, with application to DNA microarray data.
Journal of the Royal Statistical Society: Series B (Statistical Methodology) 67,
555–572.

"""
struct ConvexDecreasing <: Pi0Estimator
    gridsize::Int64
    xtol::Float64
    maxiter::Int64

    function ConvexDecreasing(gridsize, xtol, maxiter)
        gridsize > 0 || throw(DomainError("gridsize must be a positive number"))
        isin(xtol, 0, 1) || throw(DomainError("xtol must be in [0, 1]"))
        maxiter > 0 || throw(DomainError("maxiter must be a positive number"))
        return new(gridsize, xtol, maxiter)
    end
end

ConvexDecreasing() = ConvexDecreasing(100, 1e-5, 10000)

struct ConvexDecreasingFit <: Pi0Fit
    π0::Float64
    param::Vector{Float64}
    is_converged::Bool
end

function fit(pi0estimator::ConvexDecreasing, pValues::AbstractVector{T}) where T <: AbstractFloat
    π0, param, is_converged = convex_decreasing(pValues,
                                                pi0estimator.gridsize,
                                                pi0estimator.xtol,
                                                pi0estimator.maxiter)
    return ConvexDecreasingFit(π0, param, is_converged)
end

function estimate(pValues::PValues{T}, pi0estimator::ConvexDecreasing) where T <: AbstractFloat
    return estimate(fit(pi0estimator, pValues))
end

function estimate(pi0fit::ConvexDecreasingFit)
    pi0 = pi0fit.is_converged ? pi0fit.π0 : NaN
    return pi0
end

function convex_decreasing(pValues::AbstractVector{T},
                           gridsize::Integer = 100,
                           xtol::AbstractFloat = 1e-5,
                           maxiter::Integer = 10000) where T <: AbstractFloat

    n = length(pValues)
    p = sort(pValues)
    dx = 1 / gridsize
    t = collect(dx:dx:1)
    x = collect(0:dx:1)
    f = fill(one(T), gridsize + 1)
    f_p = fill(one(T), n)
    theta = dx * find_theta(t, p)
    f_theta = triangular_weighting(x, theta)
    f_theta_p = triangular_weighting(p, theta)
    idx_lower = round.(Int, round.(gridsize .* p, RoundDown) .+ 1)
    p_upper = round.(gridsize .* p, RoundUp) / gridsize
    idx_upper = round.(Int, gridsize .* p_upper) .+ 1
    px = p_upper .- p
    thetas = T[]
    pi0_new = pi0_old = Inf
    for j in 1:maxiter
        if sum((f_p .- f_theta_p) ./ f_p) > 0
            ε = 0.0
        else
            l = 0.0
            u = 1
            while abs(u - l) > xtol
                ε = (l + u) / 2.0
                if decide(f_p, f_theta_p, ε) < 0
                    l = ε
                else
                    u = ε
                end
            end
        end
        @. f = (1 - ε) * f + ε * f_theta
        @. f_p = (1 - ε) * f_p + ε * f_theta_p
        theta = dx * find_theta(t, p, f_p)
        f_theta .= triangular_weighting(x, theta)
        pi0_new = f[end]
        if abs(pi0_new - pi0_old) <= xtol
            return pi0_new, thetas, true
        end
        pi0_old = pi0_new
        f_theta_p .= triangular_weighting(p, theta)
        if sum(f_theta_p ./ f_p) < sum(1 ./ f_p)
            theta = 0.0
            f_theta .= fill(1, size(f_theta))
            f_theta_p .= fill(1, size(f_theta_p))
        end
        if !(theta in thetas)
            append!(thetas, theta)
            sort!(thetas)
        end
    end
    return NaN, thetas, false
end

function find_theta(t::Vector{Float64}, p::Vector{Float64})
    return argmax([theta.^-2 * sum(theta .- p[p .< theta]) for theta in t])
end

function find_theta(t::Vector{Float64}, p::Vector{Float64}, f_p::Vector{Float64})
    return argmax([theta.^-2 * sum((theta .- p) .* (p .< theta) ./ f_p) for theta in t])
end

function decide(f_p::Vector{Float64}, f_theta_p::Vector{Float64}, ε::Float64)
    idx = f_p .> 0
    return sum(@. ( f_p[idx] - f_theta_p[idx] ) / ( (1 - ε) * f_p[idx] + ε * f_theta_p[idx] ))
end

function triangular_weighting(x::Vector{T}, mid::T) where T <: AbstractFloat
    return @. 2 / mid^2 * (mid - x) * (x < mid)
end
